rfcs/47-additional-error-specification.md

```diff
@@ -0,0 +1,297 @@
+	case "NoSuchKey", "NotFound":
+		return fmt.Errorf("%w: %v", services.ErrObjectNotExist, err)
+	case "AccessDenied":
+		return fmt.Errorf("%w: %v", services.ErrPermissionDenied, err)
+	// What if we add a new case, but users already handle it themselves?
+	}
+
+	return err
+}
+```
+
+Although we declared in [AOS-11] that unexpected errors could be changed or disappeared while dependence upgraded and no changelog for them, we'd better simply eliminate the possibility of using them.
+
+#### Expected Errors
+
+[AOS-11] is not detailed enough. It does not specify how to define and use expected errors. Current practice has too many cases, and has some problems. Both users and implementors may be confused which kind of error should be returned at which place.
+
+- Ad-hoc string errors are not user-friendly. The user cannot use `errors.Is` & `errors.As` to handle such errors specially. We should avoid using them.
+
+- Another layer of error `struct`s may be confusing.
+  - It may be mistaken for top-level errors.
+  - Their `Err` fields are also confusing.
+    - Will they wrap yet another layer of error `struct`s? 
+    - If it will be the same for all instances, not intended for the caller to fill in, and only accessed via `Unwrap`, we can actually hide the field.
+    - The wrapped errors seem like the super class. It may be more reasonable to wrap other errors instead of being wrapped. But the order does not matter in the error chain.
+
+## Proposal
+
+So I propose the following error handling specification as a supplement of [AOS-11]:
+
+An error code is an exported public variable containing an `error` value created by `errors.New()` (a.k.a. a sentinel error)
+- `var SomeError = errors. New("what happened")`
+
+1. Public APIs SHOULD return top-level errors, which MUST be defined as below:
+	```go
+	type SomeError struct {
+		Op  string
+		Err error
+
+		ContextA string
+		ContextB structB
+		...
+	}
+	```
+   - SHOULD implement `Error` and `Unwrap`
+   - Only include `InitError`, `ServiceError` and `StorageError`
+2. For the wrapped error:
+   - If it is an unexpected error, it MUST be `fmt.Errorf("%w: %v", ErrUnexpected, err)`
+     - `ErrUnexpected` is defined as: `var ErrUnexpected = errors.New("unexpected")`. 
+     - SDK errors SHOULD not be wrapped.
+   - If it is an expected error, it MUST be either
+     - an error code
+     - an error `struct` carrying contextual information, with an assigned error code
+       - it SHOULD be defined as:
+			``` go
+			type SomeError struct {
+				ContextA string
+				ContextB structB
+				...
+			}
+			```
+       - it SHOULD implement `Unwrap`, returning its error code
+       - it SHOULD implement `Error`
+         - the format should be `{Description}, {ContextA}, {ContextB}: {Err}` , where `Err` is its error code
+         - or `{Description}, {ContextA}, {ContextB}` if it is the only struct with its error code
```

Comment:
`with` or `without`?

Reply:
`with`, I mean the error code and the struct is one-one relation, e.g. `ErrObjectModeInvalid` -  `ObjectModeInvalidError`

Now it's `object mode invalid, expected %v, actual %v` 
Otherwise it will be `object mode invalid, expected %v, actual %v: %s`
i.e.,  `object mode invalid, expected %v, actual %v: object mode invalid`

I think it's too wordy?

Reply:
And there's no struct without an error code now